### PR TITLE
Issue #227: Cache files in the same form they were on the wire

### PR DESCRIPTION
**Pull request by glandium**
Caching gzipped files in uncompressed form is a waste of disk space, so
store them in gzipped form instead. At the same properly cache on Windows
and when the downloaded file was not gzipped, which previously didn't lead
to data being cached.

Note that this means the file is gunzipped every time it's used, when it previously was read uncompressed every time, which is not necessarily better. The right thing to do would be to cache the json data in memory, but I'll leave that up to you.

### CODE REVIEW

File: mozci/utils/transfer.py [hunk 1]
```diff
@@ -121,27 +101,46 @@ def load_file(filename, url):
...
+            for chunk in req.iter_content(10 * 1024):
+                if chunk:  # filter out keep-alive new chunks
+                    fd.write(chunk)
+                    bytes += len(chunk)
+                    pbar.update(bytes)
+        pbar.finish()
+
+    elif req.status_code == 304:
+        # The file on disk is recent
+        LOG.debug("%s is on disk and it is current." % last_mod_date)
```
Comment (armenzg) [inline mozci/utils/transfer.py:141]: This message talks about file yet we subsitute with a date.

**Comment by armenzg** [comment]
Thanks glandium. My apologies about the delay. GH does not have reminders like BZ (afaik).

I've thought about it and for most cases we should not have a big performance hit. We only load a small number of those files at a time.

I would like to test this few times before I merge it in.

**Comment by armenzg** [comment]
We are fetching builds-4hr multiple times when we should not (The file only changes once every minute and we fetched it several times in few seconds).

We store the file with .js extension (even thought it is a gzip file):
-rw-r--r--  1 armenzg  staff   4283654  3 Jun 14:14 builds-4hr.js

The pastebin shows 2 fetches of builds-4hr.js.gz:
https://pastebin.mozilla.org/8835582

Here's relevant output:
06/03/2015 02:14:10 DEBUG:   We have not been able to find /Users/armenzg/.mozilla/mozci/builds-4hr.js on disk.
...
06/03/2015 02:14:17 DEBUG:   "GET /builddata/buildjson/builds-4hr.js.gz HTTP/1.1" 200 4283654
06/03/2015 02:14:17 DEBUG:   The local file was last modified at Wed, 03 Jun 2015 18:14:14 GMT. We need to fetch it again.

File: mozci/utils/transfer.py (hunk 1 above)
Comment (glandium) [inline mozci/utils/transfer.py:141]: > This message talks about file yet we subsitute with a date.

I didn't change that line ;) It comes from the current file.

**Comment by glandium** [comment]
> We store the file with .js extension (even thought it is a gzip file):

That simplified the code to do so.

> The pastebin shows 2 fetches of builds-4hr.js.gz:

The part that takes the file timestamp and requests that hasn't changed, so this wouldn't be a new problem. Can you add a log of the last-modified from the response headers?

**Comment by glandium** [comment]
That said, the time difference between the log times and the reported last modified time is weird. You're not in GMT-16 are you? ;) So there's something funky with the timezone handling, but that must have been there before.

**Comment by armenzg** [comment]
I will investigate this and merge together with any new findings.

**Comment by armenzg** [comment]
I've fixed the issue that I noticed:
https://github.com/armenzg/mozilla_ci_tools/commit/1e591bfc62e3f1afdc2a10d198208363bf666766

I will merge this in a bit.

**Comment by armenzg** [comment]
I've fixed the merge conflicts and have taken this change into "store_as_gzip" branch so we test it for few days before making it part of the main release.

Thanks @glandium !

**Comment by armenzg** [comment]
I've made some changes, tested it and pushed it.
Let's way few days before we do a release.

**Pull request merged**